File: src/server/services/impactAnalyzer.js
// Impact analysis engine for calculating market reactions to macro events

const ASSET_CATEGORIES = {
  equity: ['SPY', 'QQQ', 'IWM', 'DIA'],
  fx: ['EUR/USD', 'GBP/USD', 'USD/JPY', 'DXY'],
  bond: ['TLT', 'IEF', 'HYG'],
  volatility: ['VIX', 'VVIX'],
  commodity: ['GLD', 'USO']
};

export class ImpactAnalyzer {
  constructor() {
    this.impactThresholds = {
      minimal: 0.1,
      moderate: 0.3,
      significant: 0.5,
      major: 1.0,
      extreme: 2.0
    };
  }

  calculateImpact(beforeSnapshot, afterSnapshot, event) {
    const impacts = {};

    // Calculate impact for each asset
    Object.entries(afterSnapshot).forEach(([symbol, afterData]) => {
      const beforeData = beforeSnapshot[symbol];
      if (!beforeData) return;

      const priceChange = afterData.price - beforeData.price;
      const percentChange = (priceChange / beforeData.price) * 100;

      impacts[symbol] = {
        symbol,
        name: afterData.name,
        type: afterData.type,
        beforePrice: beforeData.price,
        afterPrice: afterData.price,
        priceChange: Math.round(priceChange * 10000) / 10000,
        percentChange: Math.round(percentChange * 100) / 100,
        magnitude: this.classifyMagnitude(Math.abs(percentChange)),
        direction: percentChange > 0 ? 'up' : percentChange < 0 ? 'down' : 'unchanged'
      };
    });

    // Calculate category aggregates
    const categoryImpacts = {};
    Object.entries(ASSET_CATEGORIES).forEach(([category, symbols]) => {
      const categoryAssets = symbols
        .filter(s => impacts[s])
        .map(s => impacts[s]);

      if (categoryAssets.length === 0) return;

      const avgChange = categoryAssets.reduce((sum, a) => sum + a.percentChange, 0) / categoryAssets.length;

      categoryImpacts[category] = {
        category,
        avgPercentChange: Math.round(avgChange * 100) / 100,
        magnitude: this.classifyMagnitude(Math.abs(avgChange)),
        direction: avgChange > 0 ? 'up' : avgChange < 0 ? 'down' : 'unchanged',
        assets: categoryAssets
      };
    });

    // Calculate overall market stress
    const overallStress = this.calculateMarketStress(impacts);

    // Determine if reaction aligns with expectations
    const expectedImpact = this.getExpectedReaction(event);
    const alignment = this.checkAlignment(categoryImpacts, expectedImpact);

    return {
      timestamp: new Date().toISOString(),
      event: {
        indicator: event.indicator,
        actual: event.actual,
        forecast: event.forecast,
        surprise: event.surprise
      },
      assetImpacts: impacts,
      categoryImpacts,
      overallStress,
      expectedImpact,
      alignment,
      summary: this.generateSummary(categoryImpacts, event)
    };
  }

  classifyMagnitude(absPercentChange) {
    if (absPercentChange >= this.impactThresholds.extreme) return 'extreme';
    if (absPercentChange >= this.impactThresholds.major) return 'major';
    if (absPercentChange >= this.impactThresholds.significant) return 'significant';
    if (absPercentChange >= this.impactThresholds.moderate) return 'moderate';
    if (absPercentChange >= this.impactThresholds.minimal) return 'minimal';
    return 'negligible';
  }

  calculateMarketStress(impacts) {
    // Use VIX change as primary stress indicator
    const vixImpact = impacts['VIX'];
    if (!vixImpact) return 'unknown';

    const vixChange = vixImpact.percentChange;

    if (vixChange > 15) return 'extreme';
    if (vixChange > 10) return 'high';
    if (vixChange > 5) return 'elevated';
    if (vixChange > 0) return 'mild';
    if (vixChange > -5) return 'calm';
    return 'very_calm';
  }

  getExpectedReaction(event) {
    const { indicator, surprise } = event;

    const expectations = {
      CPI: {
        // Higher than expected CPI = hawkish
        equity: surprise > 0 ? 'down' : 'up',
        fx: surprise > 0 ? 'up' : 'down', // USD strength
        bond: surprise > 0 ? 'down' : 'up',
        volatility: surprise > 0 ? 'up' : 'down',
        reasoning: surprise > 0
          ? 'Hot CPI suggests Fed stays hawkish, pressuring risk assets'
          : 'Cool CPI suggests Fed can ease, supporting risk assets'
      },
      NFP: {
        // Strong NFP = mixed, depends on context
        equity: surprise > 0 ? 'neutral' : 'down',
        fx: surprise > 0 ? 'up' : 'down',
        bond: surprise > 0 ? 'down' : 'up',
        volatility: Math.abs(surprise) > 20 ? 'up' : 'neutral',
        reasoning: surprise > 0
          ? 'Strong jobs support growth but keep Fed hawkish'
          : 'Weak jobs raise recession concerns'
      },
      PMI: {
        equity: surprise > 0 ? 'up' : 'down',
        fx: surprise > 0 ? 'up' : 'down',
        bond: surprise > 0 ? 'down' : 'up',
        volatility: surprise < -5 ? 'up' : 'down',
        reasoning: surprise > 0
          ? 'Expanding manufacturing supports growth outlook'
          : 'Contracting manufacturing signals economic weakness'
      },
      FOMC: {
        // For rates, negative surprise = rate cut = dovish
        equity: surprise < 0 ? 'up' : 'down',
        fx: surprise < 0 ? 'down' : 'up',
        bond: surprise < 0 ? 'up' : 'down',
        volatility: 'up', // FOMC always increases vol
        reasoning: surprise < 0
          ? 'Dovish Fed supports risk assets'
          : 'Hawkish Fed pressures valuations'
      },
      GDP: {
        equity: surprise > 0 ? 'up' : 'down',
        fx: surprise > 0 ? 'up' : 'down',
        bond: surprise > 0 ? 'down' : 'up',
        volatility: surprise < 0 ? 'up' : 'down',
        reasoning: surprise > 0
          ? 'Stronger growth supports corporate earnings'
          : 'Weaker growth raises recession concerns'
      }
    };

    return expectations[indicator] || null;
  }

  checkAlignment(categoryImpacts, expectedImpact) {
    if (!expectedImpact) return { aligned: 'unknown', score: 0 };

    let alignedCount = 0;
    let totalChecked = 0;
    const details = {};

    Object.entries(expectedImpact).forEach(([category, expected]) => {
      if (category === 'reasoning' || !categoryImpacts[category]) return;

      const actual = categoryImpacts[category].direction;
      totalChecked++;

      const isAligned = expected === 'neutral' || actual === expected || actual === 'unchanged';
      if (isAligned) alignedCount++;

      details[category] = {
        expected,
        actual,
        aligned: isAligned
      };
    });

    const score = totalChecked > 0 ? Math.round((alignedCount / totalChecked) * 100) : 0;

    return {
      aligned: score >= 75 ? 'yes' : score >= 50 ? 'partial' : 'no',
      score,
      details,
      reasoning: expectedImpact.reasoning
    };
  }

  generateSummary(categoryImpacts, event) {
    const parts = [];

    // Equity summary
    if (categoryImpacts.equity) {
      const eq = categoryImpacts.equity;
      parts.push(`Equities ${eq.direction} ${Math.abs(eq.avgPercentChange).toFixed(2)}%`);
    }

    // FX summary
    if (categoryImpacts.fx) {
      const fx = categoryImpacts.fx;
      parts.push(`USD ${fx.direction === 'up' ? 'strengthened' : 'weakened'} (DXY ${fx.direction} ${Math.abs(fx.avgPercentChange).toFixed(2)}%)`);
    }

    // Bond summary
    if (categoryImpacts.bond) {
      const bond = categoryImpacts.bond;
      parts.push(`Treasuries ${bond.direction} ${Math.abs(bond.avgPercentChange).toFixed(2)}%`);
    }

    // Volatility summary
    if (categoryImpacts.volatility) {
      const vol = categoryImpacts.volatility;
      parts.push(`VIX ${vol.direction === 'up' ? 'spiked' : 'dropped'} ${Math.abs(vol.avgPercentChange).toFixed(2)}%`);
    }

    return parts.join(' | ');
  }

  // Historical impact analysis
  analyzeHistoricalImpacts(events) {
    const analysis = {
      avgEquityMove: 0,
      avgFxMove: 0,
      avgBondMove: 0,
      avgVolMove: 0,
      surpriseCorrelation: 0,
      bestPredictors: [],
      patterns: []
    };

    if (events.length === 0) return analysis;

    // Calculate averages
    let equityMoves = [];
    let fxMoves = [];
    let bondMoves = [];
    let volMoves = [];

    events.forEach(event => {
      if (event.impacts?.['60m']?.categoryImpacts) {
        const impacts = event.impacts['60m'].categoryImpacts;
        if (impacts.equity) equityMoves.push(impacts.equity.avgPercentChange);
        if (impacts.fx) fxMoves.push(impacts.fx.avgPercentChange);
        if (impacts.bond) bondMoves.push(impacts.bond.avgPercentChange);
        if (impacts.volatility) volMoves.push(impacts.volatility.avgPercentChange);
      }
    });

    const avg = arr => arr.length > 0 ? arr.reduce((a, b) => a + b, 0) / arr.length : 0;

    analysis.avgEquityMove = Math.round(avg(equityMoves) * 100) / 100;
    analysis.avgFxMove = Math.round(avg(fxMoves) * 100) / 100;
    analysis.avgBondMove = Math.round(avg(bondMoves) * 100) / 100;
    analysis.avgVolMove = Math.round(avg(volMoves) * 100) / 100;

    return analysis;
  }
}

File: services/impact_analyzer.py
"""
Impact Analyzer - Calculate market reactions to macro events

Analyzes price changes across asset classes and determines 
the market's reaction to economic data releases.
"""

from datetime import datetime
from typing import Dict, List, Optional

# Asset categories for grouping analysis
ASSET_CATEGORIES = {
    'equity': ['SPY', 'QQQ', 'IWM', 'DIA'],
    'fx': ['EUR/USD', 'GBP/USD', 'USD/JPY', 'DXY'],
    'bond': ['TLT', 'IEF', 'HYG'],
    'volatility': ['VIX', 'VVIX'],
    'commodity': ['GLD', 'USO'],
    'crypto': ['BTC', 'ETH', 'SOL', 'XRP']
}


class ImpactAnalyzer:
    """Analyzes market impact of macro economic events."""
    
    def __init__(self):
        self.impact_thresholds = {
            'minimal': 0.1,
            'moderate': 0.3,
            'significant': 0.5,
            'major': 1.0,
            'extreme': 2.0
        }
    
    def calculate_impact(self, before_snapshot: dict, after_snapshot: dict, event: dict) -> dict:
        """Calculate the market impact between two price snapshots."""
        impacts = {}
        
        # Calculate impact for each asset
        for symbol, after_data in after_snapshot.items():
            before_data = before_snapshot.get(symbol)
            if not before_data:
                continue
            
            price_change = after_data['price'] - before_data['price']
            percent_change = (price_change / before_data['price']) * 100 if before_data['price'] else 0
            
            impacts[symbol] = {
                'symbol': symbol,
                'name': after_data['name'],
                'type': after_data['type'],
                'before_price': before_data['price'],
                'after_price': after_data['price'],
                'price_change': round(price_change, 4),
                'percent_change': round(percent_change, 2),
                'magnitude': self._classify_magnitude(abs(percent_change)),
                'direction': 'up' if percent_change > 0 else ('down' if percent_change < 0 else 'unchanged')
            }
        
        # Calculate category aggregates
        category_impacts = {}
        for category, symbols in ASSET_CATEGORIES.items():
            category_assets = [impacts[s] for s in symbols if s in impacts]
            
            if not category_assets:
                continue
            
            avg_change = sum(a['percent_change'] for a in category_assets) / len(category_assets)
            
            category_impacts[category] = {
                'category': category,
                'avg_percent_change': round(avg_change, 2),
                'magnitude': self._classify_magnitude(abs(avg_change)),
                'direction': 'up' if avg_change > 0 else ('down' if avg_change < 0 else 'unchanged'),
                'assets': category_assets
            }
        
        # Calculate overall market stress
        overall_stress = self._calculate_market_stress(impacts)
        
        # Determine if reaction aligns with expectations
        expected_impact = self._get_expected_reaction(event)
        alignment = self._check_alignment(category_impacts, expected_impact)
        
        return {
            'timestamp': datetime.utcnow().isoformat(),
            'event': {
                'indicator': event.get('indicator'),
                'actual': event.get('actual'),
                'forecast': event.get('forecast'),
                'surprise': event.get('surprise')
            },
            'asset_impacts': impacts,
            'category_impacts': category_impacts,
            'overall_stress': overall_stress,
            'expected_impact': expected_impact,
            'alignment': alignment,
            'summary': self._generate_summary(category_impacts, event)
        }
    
    def _classify_magnitude(self, abs_percent_change: float) -> str:
        """Classify the magnitude of a price change."""
        if abs_percent_change >= self.impact_thresholds['extreme']:
            return 'extreme'
        if abs_percent_change >= self.impact_thresholds['major']:
            return 'major'
        if abs_percent_change >= self.impact_thresholds['significant']:
            return 'significant'
        if abs_percent_change >= self.impact_thresholds['moderate']:
            return 'moderate'
        if abs_percent_change >= self.impact_thresholds['minimal']:
            return 'minimal'
        return 'negligible'
    
    def _calculate_market_stress(self, impacts: dict) -> str:
        """Calculate overall market stress using VIX as primary indicator."""
        vix_impact = impacts.get('VIX')
        if not vix_impact:
            return 'unknown'
        
        vix_change = vix_impact['percent_change']
        
        if vix_change > 15:
            return 'extreme'
        if vix_change > 10:
            return 'high'
        if vix_change > 5:
            return 'elevated'
        if vix_change > 0:
            return 'mild'
        if vix_change > -5:
            return 'calm'
        return 'very_calm'
    
    def _get_expected_reaction(self, event: dict) -> Optional[dict]:
        """Get expected market reaction based on indicator and surprise."""
        indicator = event.get('indicator')
        surprise = event.get('surprise', 0)
        
        expectations = {
            'CPI': {
                'equity': 'down' if surprise > 0 else 'up',
                'fx': 'up' if surprise > 0 else 'down',
                'bond': 'down' if surprise > 0 else 'up',
                'volatility': 'up' if surprise > 0 else 'down',
                'reasoning': 'Hot CPI suggests Fed stays hawkish, pressuring risk assets' if surprise > 0
                            else 'Cool CPI suggests Fed can ease, supporting risk assets'
            },
            'NFP': {
                'equity': 'neutral' if surprise > 0 else 'down',
                'fx': 'up' if surprise > 0 else 'down',
                'bond': 'down' if surprise > 0 else 'up',
                'volatility': 'up' if abs(surprise) > 20 else 'neutral',
                'reasoning': 'Strong jobs support growth but keep Fed hawkish' if surprise > 0
                            else 'Weak jobs raise recession concerns'
            },
            'PMI': {
                'equity': 'up' if surprise > 0 else 'down',
                'fx': 'up' if surprise > 0 else 'down',
                'bond': 'down' if surprise > 0 else 'up',
                'volatility': 'up' if surprise < -5 else 'down',
                'reasoning': 'Expanding manufacturing supports growth outlook' if surprise > 0
                            else 'Contracting manufacturing signals economic weakness'
            },
            'FOMC': {
                'equity': 'up' if surprise < 0 else 'down',
                'fx': 'down' if surprise < 0 else 'up',
                'bond': 'up' if surprise < 0 else 'down',
                'volatility': 'up',
                'reasoning': 'Dovish Fed supports risk assets' if surprise < 0
                            else 'Hawkish Fed pressures valuations'
            },
            'GDP': {
                'equity': 'up' if surprise > 0 else 'down',
                'fx': 'up' if surprise > 0 else 'down',
                'bond': 'down' if surprise > 0 else 'up',
                'volatility': 'up' if surprise < 0 else 'down',
                'reasoning': 'Stronger growth supports corporate earnings' if surprise > 0
                            else 'Weaker growth raises recession concerns'
            }
        }
        
        return expectations.get(indicator)
    
    def _check_alignment(self, category_impacts: dict, expected_impact: Optional[dict]) -> dict:
        """Check if actual reaction aligns with expected reaction."""
        if not expected_impact:
            return {'aligned': 'unknown', 'score': 0}
        
        aligned_count = 0
        total_checked = 0
        details = {}
        
        for category, expected in expected_impact.items():
            if category == 'reasoning' or category not in category_impacts:
                continue
            
            actual = category_impacts[category]['direction']
            total_checked += 1
            
            is_aligned = expected == 'neutral' or actual == expected or actual == 'unchanged'
            if is_aligned:
                aligned_count += 1
            
            details[category] = {
                'expected': expected,
                'actual': actual,
                'aligned': is_aligned
            }
        
        score = round((aligned_count / total_checked) * 100) if total_checked > 0 else 0
        
        return {
            'aligned': 'yes' if score >= 75 else ('partial' if score >= 50 else 'no'),
            'score': score,
            'details': details,
            'reasoning': expected_impact.get('reasoning', '')
        }
    
    def _generate_summary(self, category_impacts: dict, event: dict) -> str:
        """Generate a human-readable summary of the impact."""
        parts = []
        
        if 'equity' in category_impacts:
            eq = category_impacts['equity']
            parts.append(f"Equities {eq['direction']} {abs(eq['avg_percent_change']):.2f}%")
        
        if 'fx' in category_impacts:
            fx = category_impacts['fx']
            verb = 'strengthened' if fx['direction'] == 'up' else 'weakened'
            parts.append(f"USD {verb} (DXY {fx['direction']} {abs(fx['avg_percent_change']):.2f}%)")
        
        if 'bond' in category_impacts:
            bond = category_impacts['bond']
            parts.append(f"Treasuries {bond['direction']} {abs(bond['avg_percent_change']):.2f}%")
        
        if 'volatility' in category_impacts:
            vol = category_impacts['volatility']
            verb = 'spiked' if vol['direction'] == 'up' else 'dropped'
            parts.append(f"VIX {verb} {abs(vol['avg_percent_change']):.2f}%")
        
        return ' | '.join(parts)
    
    def analyze_historical_impacts(self, events: List[dict]) -> dict:
        """Analyze patterns from historical impact data."""
        analysis = {
            'avg_equity_move': 0,
            'avg_fx_move': 0,
            'avg_bond_move': 0,
            'avg_vol_move': 0,
            'surprise_correlation': 0,
            'best_predictors': [],
            'patterns': []
        }
        
        if not events:
            return analysis
        
        moves = {'equity': [], 'fx': [], 'bond': [], 'vol': []}
        
        for event in events:
            impacts = event.get('impacts', {}).get('60m', {}).get('category_impacts', {})
            if not impacts:
                continue
            
            if 'equity' in impacts:
                moves['equity'].append(impacts['equity']['avg_percent_change'])
            if 'fx' in impacts:
                moves['fx'].append(impacts['fx']['avg_percent_change'])
            if 'bond' in impacts:
                moves['bond'].append(impacts['bond']['avg_percent_change'])
            if 'volatility' in impacts:
                moves['vol'].append(impacts['volatility']['avg_percent_change'])
        
        def avg(arr):
            return sum(arr) / len(arr) if arr else 0
        
        analysis['avg_equity_move'] = round(avg(moves['equity']), 2)
        analysis['avg_fx_move'] = round(avg(moves['fx']), 2)
        analysis['avg_bond_move'] = round(avg(moves['bond']), 2)
        analysis['avg_vol_move'] = round(avg(moves['vol']), 2)
        
        return analysis
